paddle/fluid/operators/lrn_mkldnn_op.cc

```diff
@@ -0,0 +1,178 @@
+ public:
+  void Compute(const paddle::framework::ExecutionContext& ctx) const override {
+    PADDLE_ENFORCE(std::is_same<T, float>::value,
+                   "MKLDNN LRN must use float data.");
+    PADDLE_ENFORCE(paddle::platform::is_cpu_place(ctx.GetPlace()),
+                   "MKLDNN LRN must use CPUPlace.");
+
+    auto& dev_ctx = ctx.template device_context<MKLDNNDeviceContext>();
+    const auto& mkldnn_engine = dev_ctx.GetEngine();
+
+    auto x = ctx.Input<Tensor>("X");
+    auto out = ctx.Output<Tensor>("Out");
+    auto mid = ctx.Output<Tensor>("MidOut");
+
+    auto input_data = x->data<T>();
+    auto output_data = out->mutable_data<T>(ctx.GetPlace());
+    mid->mutable_data<T>(ctx.GetPlace());
+
+    const std::string key = ctx.op().Output("Out");
+    const std::string key_src_memory = key + "@lrn_src_memory";
+    const std::string key_pd = key + "@lrn_pd";
+    const std::string key_workspace_memory = key + "@lrn_workspace_memory";
+
+    const int n = ctx.Attr<int>("n");
+    const float alpha = ctx.Attr<float>("alpha");
+    const float beta = ctx.Attr<float>("beta");
+    const float k = ctx.Attr<float>("k");
+
+    auto e_mid = framework::EigenTensor<T, 4>::From(*mid);
+    e_mid = e_mid.constant(k);
+
+    auto dims = paddle::framework::vectorize2int(x->dims());
+
+    auto src_md = paddle::platform::MKLDNNMemDesc(
+        dims, mkldnn::memory::data_type::f32, mkldnn::memory::format::nchw);
+
+    auto dst_md = paddle::platform::MKLDNNMemDesc(
+        dims, mkldnn::memory::data_type::f32, mkldnn::memory::format::nchw);
+
+    auto forward_desc = mkldnn::lrn_forward::desc{mkldnn::prop_kind::forward,
+                                                  mkldnn::lrn_across_channels,
+                                                  src_md,
+                                                  n,
+                                                  alpha,
+                                                  beta,
+                                                  k};
+
+    auto forward_pd = std::make_shared<mkldnn::lrn_forward::primitive_desc>(
+        forward_desc, mkldnn_engine);
+
+    dev_ctx.SetBlob(key_pd, forward_pd);
+
+    auto src_memory_pd = mkldnn::memory::primitive_desc{src_md, mkldnn_engine};
+    auto src_memory = std::make_shared<mkldnn::memory>(
+        src_memory_pd, static_cast<void*>(const_cast<float*>(input_data)));
+
+    dev_ctx.SetBlob(key_src_memory, src_memory);
+    auto dst_memory = mkldnn::memory{{dst_md, mkldnn_engine},
+                                     static_cast<void*>(output_data)};
+
+    auto workspace_md = forward_pd->workspace_primitive_desc();
+    auto workspace_memory = std::make_shared<mkldnn::memory>(workspace_md);
+
+    dev_ctx.SetBlob(key_workspace_memory, workspace_memory);
+
+    auto forward_op = mkldnn::lrn_forward{*forward_pd, *src_memory,
+                                          *workspace_memory, dst_memory};
```

Comment:
BTW, this memory and function is used only in training phase,  maybe you should add one TODO to optimize it when you can get phase, as a reminder.

Reply:
@tensor-tang, good point! Do you know how to discover how to check whether we are in training phase or scoring phase?

Reply:
You can following the same way of batch_norm_op, which uses `is_test`:
https://github.com/PaddlePaddle/Paddle/blob/0821ee78c5c553ff5ffe1fde40a73c01cf325b5e/paddle/fluid/operators/batch_norm_op.cc#L175-L182

Reply:
@luotao1, I do see your point but in case of LRN `is_test` would become a part of user interface not only for MKLDNN LRN, but also for other implementations of LRN. Do you think adding such attribute would be justified for other implementations of LRN operator?

You could make such case for other operators as well. We could end up in the situation when each operator sets its own local `is_test` attribute to maintain information about phase, when this information is in fact global (global as being maintain on the network level).

Would it make more sense to maintain phase information in network executor?

I could add `TODO` comment saying that the blob should not be allocated when PaddlePaddle is in testing phase.

Reply:
> Do you think adding such attribute would be justified for other implementations of LRN operator

Yes, "is_test" is also suited for other implementation of LRN operators. 

> You could make such case for other operators as well. We could end up in the situation when each operator sets its own local is_test attribute to maintain information about phase, when this information is in fact global (global as being maintain on the network level). Would it make more sense to maintain phase information in network executor?

Now, `framework/prune.cc` use this attribute in inference phase. 
https://github.com/PaddlePaddle/Paddle/blob/0821ee78c5c553ff5ffe1fde40a73c01cf325b5e/paddle/fluid/framework/prune.cc#L187-L197

> I could add TODO comment saying that the blob should not be allocated when PaddlePaddle is in testing phase.

Yes, you can either add TODO comment or fix in this PR.

Reply:
@luotao1 thanks for your reply. I will add the attribute to the operator. 

But would it be more useful to maintain testing/training information on the network level, instead of keeping on the operator level?

Reply:
Now We maintain testing/training information on network level as well.  https://github.com/PaddlePaddle/Paddle/blob/0821ee78c5c553ff5ffe1fde40a73c01cf325b5e/python/paddle/fluid/framework.py#L959-L982

But if operator level doesn't have is_test information, how could this op behave differently in test or train phase?